### ISSUE TITLE
feat: Add support of ODCS (Open Data Contract Standard)

### ISSUE THREAD

**Issue by Linux-oiD** [datahub-project/datahub#11892]
Hey team!
It would be nice if DataHub can import data contract / asserts definition from the [ODCS yaml files](https://bitol-io.github.io/open-data-contract-standard/latest/).
This project is supported by Linux Foundation and rapidly grows its popularity in the data community.
For example developers of alternative solution datacontract.com have added support of ODCS v3.0 to their [CLI tool](https://github.com/datacontract/datacontract-cli/releases/tag/v0.10.14) and in long-term are going to merge both standards.

**Comment by taabrcr** [comment]
I would find this as well very useful. Any input on this one @RyanHolstien @mhw @rubiojr @cburroughs?

**Comment by askumar27** [comment]
Attaching another request for this feature from slack: https://datahubspace.slack.com/archives/CUMUWQU66/p1766053459152979 

cc: @yoonhyejin  @deepgarg760

**Comment by Linux-oiD** [comment]
Team!
Any updates / plans to take it in work?

**Comment by kzajaczkowski** [comment]
Hi Team, that would be indeed a great feature! Any update on it?

**Comment by rbrenton** [comment]
Opened a pair of draft PRs implementing ODCS v3.0 / v3.1 ingestion:

- **#17331** - `feat(ingest/odcs): add Open Data Contract Standard ingestion source` - the new `source.type: odcs` plugin
- **#17332** - `feat(metadata-models): register odcs as a DataPlatform` - companion bootstrap-MCP registration, split out so the platform id (`odcs`) and display name ("Open Data Contract Standard") can be reviewed independently

What the MVP supports:

- Multi-table contracts fan out to one DataHub Dataset URN per `schema[]` entry (the upstream Bitol AdventureWorks example has 4 tables; this works without cross-table metadata bleed)
- Per-table emission of dataset description, ownership, top-level + column-level tags, editable schema metadata, and assertions from `quality[]` rules (`unique`/`notNull`/`rowCount` library and SQL rules with `query`; everything else, including `mustNotBeBetween`, routes to `CustomAssertionInfo` with explicit `logic`)
- Original YAML preserved verbatim in `DataContractProperties.rawContract` (1 MB default cap)
- Vendored ODCS JSON Schemas pinned to upstream release tags v3.0.2 / v3.1.0 with SHA-256 checksums and a refresh script that refuses to overwrite on hash mismatch
- Soft-delete scoped to ODCS-owned `DataContract` and `Assertion` URNs only - Datasets are owned by their platform-of-record source (postgres / snowflake / etc.) and are never marked removed by ODCS
- Apache 2.0 attribution added to `/NOTICE`; lenient JSON-Schema validation by default with structured warnings on unknown fields

Deliberately deferred to follow-ups: SLA / pricing / support / non-owner team roles, classification → GlossaryTerm linking, ODCS export (DataHub → ODCS), ODCS v2.x support, optional `DataProductProperties` aspect for grouped contracts, per-table mixed-platform `servers[]` mapping, and per-table `dataset_urn_overrides`.

Both PRs are in draft so reviewers can preview the scope before formal review. Feedback welcome on naming, configuration shape, or any of the deferred items - particularly which followups would unblock specific adopters.

**Comment by AdrianMachado** [comment]
Will try and get PRs reviewed this week

**Comment by rbrenton** [comment]
**Update (June):** Following review on #17331, the model pivoted from emitting `dataContract` entities to **logical datasets**, which fits ODCS better (it's a dataset spec/interface, not a bilateral agreement). Net changes vs. the original scope above:

- ODCS now ingests each `schema[]` entry as a **logical `dataset`** on the `odcs` platform. When a physical table is bound (via `physical_urn_overrides`), the physical dataset gets a `logicalParent` (PhysicalInstanceOf) link; assertions are emitted only against a resolved physical binding.
- **Soft-delete scope** is now the logical `odcs` Datasets + Assertions ODCS owns; physical datasets and their `logicalParent` links are never marked removed.
- `DataContractProperties.rawContract` (and its size cap) is **removed** — ODCS no longer writes `dataContract` aspects.
- `dataset_urn_overrides` → `physical_urn_overrides` (a list per contract, matching the `schema[]` fan-out).
- #17332 also registers the `odcs` platform display name + logo.

Both PRs are green and out of draft, pending final review.

**Comment by AdrianMachado** [comment]
The new ODCS Ingestion source will be arriving in the next release of DataHub! Its been a long journey, but I am happy we can deliver. You can read through [the docs](https://docs.datahub.com/docs/generated/ingestion/sources/odcs) for more details on how it works and how concepts map to DataHub.
